src/cat/cat.rs

```diff
@@ -276,18 +289,33 @@ fn open(path: &str) -> CatResult<InputHandle> {
 /// * `files` - There is no short circuit when encountiner an error
 /// reading a file in this vector
 fn write_fast(files: Vec<String>) -> CatResult<()> {
-    let mut writer = stdout();
-    let mut in_buf = [0; 1024 * 64];
     let mut error_count = 0;
+    let writer = stdout();
+    let mut writer_handle = writer.lock();
 
     for file in files {
         match open(&file[..]) {
-            Ok(mut handle) => while let Ok(n) = handle.reader.read(&mut in_buf) {
-                if n == 0 {
-                    break;
+            Ok(mut handle) => {
+                // If we're on Linux or Android, try to use the splice()
+                // system call for faster writing.
+                #[cfg(any(target_os = "linux", target_os = "android"))]
+                {
+                    match write_fast_using_splice(&mut handle, writer.as_raw_fd()) {
+                        Ok(_) => {
+                            // Writing fast with splice worked! We don't need
+                            // to fall back on slower writing.
+                            continue;
+                        }
+                        Err(_) => {
+                            // Ignore any error and fall back to slower
+                            // writing below.
+                        }
+                    }
                 }
-                writer.write_all(&in_buf[..n]).context(&file[..])?;
-            },
+                // If we're not on Linux or Android, or the splice() call failed,
+                // fall back on slower writing.
+                io::copy(&mut handle.reader, &mut writer_handle).unwrap();
```

Comment:
The old implementation was actually incorrect in that this should not panic/cause an error.  It should write out the error message and then bump the error count like in the `Err(error)` clause below.